In [1]:
from numpy import *
import numpy as np
import pandas as pd
import pylab as pl
import matplotlib.pyplot as plt
from numpy import linalg as la
import pdb
%matplotlib inline

In [2]:
rr = pd.read_json('renttherunway_final_data.json',lines=True)
print(shape(rr))
rr.head(5)

category_list = ['gown','jacket']

rr = rr[rr['category'].isin(category_list)]

(192544, 15)


In [3]:
rr.head()

fit  user_id bust size  item_id  weight  rating     rented for  \
1   fit   273551       34b   153475  132lbs    10.0          other   
4   fit   151944       34b   616682  145lbs    10.0        wedding   
8   fit   166228       36d  1729232     NaN    10.0  formal affair   
9   fit   154309       32b  1729232  114lbs    10.0  formal affair   
14  fit   721308       34b   123793  118lbs    10.0  formal affair   

                                          review_text          body type  \
1   I rented this dress for a photo shoot. The the...  straight & narrow   
4   I have always been petite in my upper body and...           athletic   
8   I was nervous of it looking cheap when it arri...          full bust   
9   The dress was very flattering and fit perfectl...             petite   
14                        Fit great, super flattering           athletic   

                                       review_summary category height  size  \
1                             I felt so glamourous!!!     gown  5' 6"    12   
4                     Was in love with this dress !!!     gown  5' 9"    12   
8                          Great for black tie event!     gown  5' 6"    21   
9   This dress was everything! It was perfect for ...     gown  5' 3"     1   
14  Stunning gown. Wore this for heart ball and re...     gown  5' 5"     2   

     age         review_date  
1   36.0       June 18, 2013  
4   27.0  September 26, 2016  
8   27.0       June 27, 2016  
9   33.0    October 17, 2016  
14  32.0        May 29, 2014

In [4]:
#reorder id's
item_id = rr['item_id']
user_id = rr['user_id']
rating = rr['rating']

rr.drop('item_id', axis = 1,inplace = True)
rr.insert(0,'item_id',item_id)
rr.drop('user_id', axis = 1, inplace  = True)
rr.insert(1,'user_id',user_id)
rr.drop('rating', axis = 1,inplace = True)
rr.insert(2,'rating',rating)

print(shape(rr))


(46785, 15)


In [7]:
rr_Sorted = rr.sort_values(['user_id'])
n_rr = rr_Sorted.drop_duplicates(['user_id'])
shape(n_rr)

n_rr_Sorted = n_rr.sort_values(['item_id'])
n_rr = n_rr_Sorted.drop_duplicates(['item_id'])
shape(n_rr)

n_rr_rating = n_rr.iloc[:,0:3]

n_rr_rating.item_id = n_rr_rating.item_id.astype(object)
n_rr_rating.user_id = n_rr_rating.user_id.astype(object)

##### User x Item Matrix

In [8]:
# user x item
R = n_rr_rating.pivot_table(index = ['user_id'],columns = ['item_id'],values = "rating")
R1 = R.replace(np.nan,0)
print(shape(R1))
print(R1.head())
R1 = np.array(R1)

(849, 849)
item_id  123373   123793   125465   125564   127081   127865   128959   \
user_id                                                                  
35           0.0      0.0      0.0      0.0      0.0      0.0      0.0   
2132         0.0      0.0      0.0      0.0      0.0      0.0      0.0   
2968         0.0      0.0      0.0      0.0      0.0      0.0      0.0   
4170         0.0      0.0      0.0      0.0      0.0      0.0      0.0   
5790         0.0      0.0      0.0      0.0      0.0      0.0      0.0   

item_id  131117   131533   131698   ...  2935830  2935879  2941989  2945782  \
user_id                             ...                                       
35           0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
2132         0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
2968         0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
4170         0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0

##### Matrix Factorization

The matrix factorization function below accepts a matrix R, which is the User x Item matrix or the Rating Matrix. It then creates matrix P and matrix Q using stochasitic gradient descent. Matrix P (User Matrix) is the User x Factors matrix and matrix Q (Item Matrix) is the Item x factor matrix. K refers to the number of features, steps are the number of ephocs per gradient, alpha is the learning rate, and beta is the regularization coefficient. 

eij, or the rating, is calculated with the dot product of P(i) and Q(j). P represents how much the user likes a certain latent factor and Q refers to the how related an item is to a latent factor. 

In [9]:
def matrix_factorization(R, P, Q, K, steps=5000, alpha=0.0002, beta=0.02):
    
    ### R = The user x item rating matrix (m x n)
    ### P = Initial user-factor matrix (m x k)
    ### Q = Initial item-factor matrix (n x k)
    ### K = The number of latent factors (features)
    ### steps = The number of epochs in gradient descent
    ### alpha = The learning rate for gradient descent
    ### beta = The regularization coefficient
    
    Q = Q.T
    for step in range(steps):
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    eij = R[i][j] - np.dot(P[i,:],Q[:,j])
                    for k in range(K):
                        ### update P and Q based on the partial derivatives
                        P[i][k] = P[i][k] + alpha * (2 * eij * Q[k][j] - beta * P[i][k])
                        Q[k][j] = Q[k][j] + alpha * (2 * eij * P[i][k] - beta * Q[k][j])
        eR = np.dot(P,Q)
        e = 0
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    e = e + pow(R[i][j] - np.dot(P[i,:],Q[:,j]), 2)
                    for k in range(K):
                        e = e + (beta/2) * ( pow(P[i][k],2) + pow(Q[k][j],2) )
        if e < 0.001:
            break
        print("Step %d of %d; Error: %0.5f; Time: %0.2f" %(step+1, steps, e, time()))
    return P, Q.T

##### P and Q
P and Q are initialized using the dimensions of the Ratings matrix (M x N). Matrix factorization is applied and P and Q are saved.

In [10]:
M = R1.shape[0]
N = R1.shape[1]
K = 12
P = np.random.rand(M,K)
Q = np.random.rand(N,K)

from time import time
t0 = time()

fP, fQ = matrix_factorization(R1, P, Q, 12, steps=3000, alpha=0.0002, beta=0.02)
print("done in %0.3fs." % (time() - t0))


Step 1 of 3000; Error: 34164.91999; Time: 1605824996.99
Step 2 of 3000; Error: 33952.90912; Time: 1605824997.98
Step 3 of 3000; Error: 33741.44503; Time: 1605824998.94
Step 4 of 3000; Error: 33530.53096; Time: 1605825000.18
Step 5 of 3000; Error: 33320.17020; Time: 1605825001.45
Step 6 of 3000; Error: 33110.36606; Time: 1605825002.69
Step 7 of 3000; Error: 32901.12187; Time: 1605825004.08
Step 8 of 3000; Error: 32692.44100; Time: 1605825005.64
Step 9 of 3000; Error: 32484.32683; Time: 1605825007.12
Step 10 of 3000; Error: 32276.78279; Time: 1605825008.50
Step 11 of 3000; Error: 32069.81232; Time: 1605825009.99
Step 12 of 3000; Error: 31863.41886; Time: 1605825011.25
Step 13 of 3000; Error: 31657.60592; Time: 1605825012.79
Step 14 of 3000; Error: 31452.37697; Time: 1605825014.23
Step 15 of 3000; Error: 31247.73555; Time: 1605825015.71
Step 16 of 3000; Error: 31043.68517; Time: 1605825016.93
Step 17 of 3000; Error: 30840.22939; Time: 1605825018.10
Step 18 of 3000; Error: 30637.37175; Tim

Step 145 of 3000; Error: 10762.77214; Time: 1605825172.65
Step 146 of 3000; Error: 10657.02608; Time: 1605825173.63
Step 147 of 3000; Error: 10552.06654; Time: 1605825174.64
Step 148 of 3000; Error: 10447.89158; Time: 1605825175.62
Step 149 of 3000; Error: 10344.49918; Time: 1605825176.57
Step 150 of 3000; Error: 10241.88729; Time: 1605825177.53
Step 151 of 3000; Error: 10140.05380; Time: 1605825178.47
Step 152 of 3000; Error: 10038.99659; Time: 1605825179.40
Step 153 of 3000; Error: 9938.71344; Time: 1605825180.30
Step 154 of 3000; Error: 9839.20212; Time: 1605825181.26
Step 155 of 3000; Error: 9740.46035; Time: 1605825182.21
Step 156 of 3000; Error: 9642.48579; Time: 1605825183.24
Step 157 of 3000; Error: 9545.27608; Time: 1605825184.39
Step 158 of 3000; Error: 9448.82880; Time: 1605825186.02
Step 159 of 3000; Error: 9353.14148; Time: 1605825187.17
Step 160 of 3000; Error: 9258.21164; Time: 1605825188.41
Step 161 of 3000; Error: 9164.03672; Time: 1605825189.68
Step 162 of 3000; Error

Step 289 of 3000; Error: 2147.67847; Time: 1605825336.20
Step 290 of 3000; Error: 2122.10287; Time: 1605825337.52
Step 291 of 3000; Error: 2096.82990; Time: 1605825338.72
Step 292 of 3000; Error: 2071.85653; Time: 1605825339.91
Step 293 of 3000; Error: 2047.17976; Time: 1605825341.00
Step 294 of 3000; Error: 2022.79659; Time: 1605825342.14
Step 295 of 3000; Error: 1998.70406; Time: 1605825343.40
Step 296 of 3000; Error: 1974.89922; Time: 1605825344.52
Step 297 of 3000; Error: 1951.37914; Time: 1605825345.69
Step 298 of 3000; Error: 1928.14092; Time: 1605825346.73
Step 299 of 3000; Error: 1905.18166; Time: 1605825347.79
Step 300 of 3000; Error: 1882.49851; Time: 1605825348.81
Step 301 of 3000; Error: 1860.08860; Time: 1605825349.82
Step 302 of 3000; Error: 1837.94912; Time: 1605825350.82
Step 303 of 3000; Error: 1816.07725; Time: 1605825351.86
Step 304 of 3000; Error: 1794.47020; Time: 1605825353.00
Step 305 of 3000; Error: 1773.12522; Time: 1605825354.12
Step 306 of 3000; Error: 1752.0

Step 435 of 3000; Error: 432.21572; Time: 1605825510.82
Step 436 of 3000; Error: 428.43726; Time: 1605825512.02
Step 437 of 3000; Error: 424.71105; Time: 1605825513.35
Step 438 of 3000; Error: 421.03640; Time: 1605825514.68
Step 439 of 3000; Error: 417.41262; Time: 1605825516.12
Step 440 of 3000; Error: 413.83901; Time: 1605825517.39
Step 441 of 3000; Error: 410.31492; Time: 1605825518.63
Step 442 of 3000; Error: 406.83968; Time: 1605825519.97
Step 443 of 3000; Error: 403.41263; Time: 1605825521.37
Step 444 of 3000; Error: 400.03312; Time: 1605825522.79
Step 445 of 3000; Error: 396.70052; Time: 1605825523.98
Step 446 of 3000; Error: 393.41419; Time: 1605825525.09
Step 447 of 3000; Error: 390.17351; Time: 1605825526.15
Step 448 of 3000; Error: 386.97787; Time: 1605825527.18
Step 449 of 3000; Error: 383.82666; Time: 1605825528.24
Step 450 of 3000; Error: 380.71927; Time: 1605825529.27
Step 451 of 3000; Error: 377.65512; Time: 1605825530.30
Step 452 of 3000; Error: 374.63362; Time: 160582

Step 582 of 3000; Error: 196.30718; Time: 1605825680.69
Step 583 of 3000; Error: 195.83714; Time: 1605825681.85
Step 584 of 3000; Error: 195.37374; Time: 1605825682.89
Step 585 of 3000; Error: 194.91688; Time: 1605825683.91
Step 586 of 3000; Error: 194.46649; Time: 1605825684.92
Step 587 of 3000; Error: 194.02246; Time: 1605825685.93
Step 588 of 3000; Error: 193.58470; Time: 1605825686.95
Step 589 of 3000; Error: 193.15313; Time: 1605825687.99
Step 590 of 3000; Error: 192.72766; Time: 1605825689.04
Step 591 of 3000; Error: 192.30819; Time: 1605825690.07
Step 592 of 3000; Error: 191.89465; Time: 1605825691.08
Step 593 of 3000; Error: 191.48695; Time: 1605825692.13
Step 594 of 3000; Error: 191.08500; Time: 1605825693.18
Step 595 of 3000; Error: 190.68873; Time: 1605825694.20
Step 596 of 3000; Error: 190.29805; Time: 1605825695.25
Step 597 of 3000; Error: 189.91289; Time: 1605825696.29
Step 598 of 3000; Error: 189.53316; Time: 1605825697.43
Step 599 of 3000; Error: 189.15879; Time: 160582

Step 729 of 3000; Error: 167.01438; Time: 1605825839.75
Step 730 of 3000; Error: 166.95448; Time: 1605825840.76
Step 731 of 3000; Error: 166.89541; Time: 1605825841.87
Step 732 of 3000; Error: 166.83713; Time: 1605825842.88
Step 733 of 3000; Error: 166.77965; Time: 1605825843.86
Step 734 of 3000; Error: 166.72294; Time: 1605825844.85
Step 735 of 3000; Error: 166.66701; Time: 1605825845.83
Step 736 of 3000; Error: 166.61183; Time: 1605825846.83
Step 737 of 3000; Error: 166.55741; Time: 1605825847.82
Step 738 of 3000; Error: 166.50372; Time: 1605825848.79
Step 739 of 3000; Error: 166.45076; Time: 1605825849.73
Step 740 of 3000; Error: 166.39851; Time: 1605825850.70
Step 741 of 3000; Error: 166.34698; Time: 1605825851.68
Step 742 of 3000; Error: 166.29614; Time: 1605825852.64
Step 743 of 3000; Error: 166.24598; Time: 1605825853.61
Step 744 of 3000; Error: 166.19651; Time: 1605825854.84
Step 745 of 3000; Error: 166.14771; Time: 1605825856.02
Step 746 of 3000; Error: 166.09956; Time: 160582

Step 876 of 3000; Error: 163.14892; Time: 1605825982.41
Step 877 of 3000; Error: 163.14030; Time: 1605825983.51
Step 878 of 3000; Error: 163.13180; Time: 1605825984.73
Step 879 of 3000; Error: 163.12339; Time: 1605825986.03
Step 880 of 3000; Error: 163.11510; Time: 1605825987.17
Step 881 of 3000; Error: 163.10690; Time: 1605825988.29
Step 882 of 3000; Error: 163.09881; Time: 1605825989.44
Step 883 of 3000; Error: 163.09082; Time: 1605825990.63
Step 884 of 3000; Error: 163.08293; Time: 1605825991.76
Step 885 of 3000; Error: 163.07514; Time: 1605825992.89
Step 886 of 3000; Error: 163.06744; Time: 1605825994.00
Step 887 of 3000; Error: 163.05985; Time: 1605825995.15
Step 888 of 3000; Error: 163.05234; Time: 1605825996.29
Step 889 of 3000; Error: 163.04493; Time: 1605825997.39
Step 890 of 3000; Error: 163.03761; Time: 1605825998.56
Step 891 of 3000; Error: 163.03038; Time: 1605825999.70
Step 892 of 3000; Error: 163.02324; Time: 1605826000.88
Step 893 of 3000; Error: 163.01619; Time: 160582

Step 1023 of 3000; Error: 162.55563; Time: 1605826145.75
Step 1024 of 3000; Error: 162.55411; Time: 1605826146.68
Step 1025 of 3000; Error: 162.55260; Time: 1605826147.60
Step 1026 of 3000; Error: 162.55111; Time: 1605826148.54
Step 1027 of 3000; Error: 162.54963; Time: 1605826149.46
Step 1028 of 3000; Error: 162.54817; Time: 1605826150.39
Step 1029 of 3000; Error: 162.54672; Time: 1605826151.32
Step 1030 of 3000; Error: 162.54529; Time: 1605826152.26
Step 1031 of 3000; Error: 162.54388; Time: 1605826153.21
Step 1032 of 3000; Error: 162.54248; Time: 1605826154.13
Step 1033 of 3000; Error: 162.54109; Time: 1605826155.05
Step 1034 of 3000; Error: 162.53972; Time: 1605826155.97
Step 1035 of 3000; Error: 162.53837; Time: 1605826156.89
Step 1036 of 3000; Error: 162.53703; Time: 1605826157.82
Step 1037 of 3000; Error: 162.53570; Time: 1605826158.72
Step 1038 of 3000; Error: 162.53439; Time: 1605826159.63
Step 1039 of 3000; Error: 162.53309; Time: 1605826160.53
Step 1040 of 3000; Error: 162.5

Step 1167 of 3000; Error: 162.43887; Time: 1605826306.67
Step 1168 of 3000; Error: 162.43847; Time: 1605826307.84
Step 1169 of 3000; Error: 162.43808; Time: 1605826308.91
Step 1170 of 3000; Error: 162.43769; Time: 1605826309.96
Step 1171 of 3000; Error: 162.43731; Time: 1605826311.06
Step 1172 of 3000; Error: 162.43693; Time: 1605826312.14
Step 1173 of 3000; Error: 162.43655; Time: 1605826313.19
Step 1174 of 3000; Error: 162.43617; Time: 1605826314.33
Step 1175 of 3000; Error: 162.43580; Time: 1605826315.40
Step 1176 of 3000; Error: 162.43543; Time: 1605826316.57
Step 1177 of 3000; Error: 162.43506; Time: 1605826317.68
Step 1178 of 3000; Error: 162.43469; Time: 1605826318.74
Step 1179 of 3000; Error: 162.43433; Time: 1605826319.78
Step 1180 of 3000; Error: 162.43397; Time: 1605826320.91
Step 1181 of 3000; Error: 162.43362; Time: 1605826322.21
Step 1182 of 3000; Error: 162.43326; Time: 1605826323.52
Step 1183 of 3000; Error: 162.43291; Time: 1605826324.84
Step 1184 of 3000; Error: 162.4

Step 1312 of 3000; Error: 162.40186; Time: 1605826488.78
Step 1313 of 3000; Error: 162.40169; Time: 1605826489.88
Step 1314 of 3000; Error: 162.40152; Time: 1605826491.03
Step 1315 of 3000; Error: 162.40135; Time: 1605826492.19
Step 1316 of 3000; Error: 162.40118; Time: 1605826493.30
Step 1317 of 3000; Error: 162.40101; Time: 1605826494.40
Step 1318 of 3000; Error: 162.40084; Time: 1605826495.39
Step 1319 of 3000; Error: 162.40068; Time: 1605826496.66
Step 1320 of 3000; Error: 162.40051; Time: 1605826497.98
Step 1321 of 3000; Error: 162.40035; Time: 1605826499.15
Step 1322 of 3000; Error: 162.40018; Time: 1605826500.30
Step 1323 of 3000; Error: 162.40002; Time: 1605826501.44
Step 1324 of 3000; Error: 162.39986; Time: 1605826502.58
Step 1325 of 3000; Error: 162.39969; Time: 1605826503.71
Step 1326 of 3000; Error: 162.39953; Time: 1605826504.88
Step 1327 of 3000; Error: 162.39937; Time: 1605826506.03
Step 1328 of 3000; Error: 162.39921; Time: 1605826507.22
Step 1329 of 3000; Error: 162.3

Step 1456 of 3000; Error: 162.38187; Time: 1605826656.76
Step 1457 of 3000; Error: 162.38175; Time: 1605826658.07
Step 1458 of 3000; Error: 162.38163; Time: 1605826659.33
Step 1459 of 3000; Error: 162.38151; Time: 1605826660.60
Step 1460 of 3000; Error: 162.38139; Time: 1605826661.66
Step 1461 of 3000; Error: 162.38128; Time: 1605826662.78
Step 1462 of 3000; Error: 162.38116; Time: 1605826663.99
Step 1463 of 3000; Error: 162.38104; Time: 1605826665.01
Step 1464 of 3000; Error: 162.38092; Time: 1605826666.11
Step 1465 of 3000; Error: 162.38081; Time: 1605826667.18
Step 1466 of 3000; Error: 162.38069; Time: 1605826668.31
Step 1467 of 3000; Error: 162.38057; Time: 1605826669.35
Step 1468 of 3000; Error: 162.38045; Time: 1605826670.39
Step 1469 of 3000; Error: 162.38034; Time: 1605826671.44
Step 1470 of 3000; Error: 162.38022; Time: 1605826672.50
Step 1471 of 3000; Error: 162.38011; Time: 1605826673.72
Step 1472 of 3000; Error: 162.37999; Time: 1605826674.92
Step 1473 of 3000; Error: 162.3

Step 1600 of 3000; Error: 162.36600; Time: 1605826842.81
Step 1601 of 3000; Error: 162.36590; Time: 1605826843.95
Step 1602 of 3000; Error: 162.36579; Time: 1605826845.10
Step 1603 of 3000; Error: 162.36569; Time: 1605826846.29
Step 1604 of 3000; Error: 162.36558; Time: 1605826847.40
Step 1605 of 3000; Error: 162.36548; Time: 1605826848.52
Step 1606 of 3000; Error: 162.36537; Time: 1605826849.79
Step 1607 of 3000; Error: 162.36527; Time: 1605826851.03
Step 1608 of 3000; Error: 162.36517; Time: 1605826852.10
Step 1609 of 3000; Error: 162.36506; Time: 1605826853.18
Step 1610 of 3000; Error: 162.36496; Time: 1605826854.31
Step 1611 of 3000; Error: 162.36485; Time: 1605826855.38
Step 1612 of 3000; Error: 162.36475; Time: 1605826856.32
Step 1613 of 3000; Error: 162.36465; Time: 1605826857.36
Step 1614 of 3000; Error: 162.36454; Time: 1605826858.35
Step 1615 of 3000; Error: 162.36444; Time: 1605826859.28
Step 1616 of 3000; Error: 162.36433; Time: 1605826860.23
Step 1617 of 3000; Error: 162.3

Step 1744 of 3000; Error: 162.35132; Time: 1605827002.33
Step 1745 of 3000; Error: 162.35122; Time: 1605827003.40
Step 1746 of 3000; Error: 162.35112; Time: 1605827004.51
Step 1747 of 3000; Error: 162.35102; Time: 1605827005.62
Step 1748 of 3000; Error: 162.35092; Time: 1605827006.87
Step 1749 of 3000; Error: 162.35082; Time: 1605827008.13
Step 1750 of 3000; Error: 162.35072; Time: 1605827009.40
Step 1751 of 3000; Error: 162.35062; Time: 1605827010.83
Step 1752 of 3000; Error: 162.35052; Time: 1605827012.14
Step 1753 of 3000; Error: 162.35042; Time: 1605827013.42
Step 1754 of 3000; Error: 162.35032; Time: 1605827014.67
Step 1755 of 3000; Error: 162.35022; Time: 1605827015.90
Step 1756 of 3000; Error: 162.35012; Time: 1605827017.30
Step 1757 of 3000; Error: 162.35002; Time: 1605827018.58
Step 1758 of 3000; Error: 162.34992; Time: 1605827019.82
Step 1759 of 3000; Error: 162.34982; Time: 1605827021.02
Step 1760 of 3000; Error: 162.34972; Time: 1605827022.18
Step 1761 of 3000; Error: 162.3

Step 1888 of 3000; Error: 162.33704; Time: 1605827168.28
Step 1889 of 3000; Error: 162.33694; Time: 1605827169.28
Step 1890 of 3000; Error: 162.33684; Time: 1605827170.43
Step 1891 of 3000; Error: 162.33674; Time: 1605827171.42
Step 1892 of 3000; Error: 162.33665; Time: 1605827172.46
Step 1893 of 3000; Error: 162.33655; Time: 1605827173.61
Step 1894 of 3000; Error: 162.33645; Time: 1605827175.06
Step 1895 of 3000; Error: 162.33635; Time: 1605827176.29
Step 1896 of 3000; Error: 162.33625; Time: 1605827177.53
Step 1897 of 3000; Error: 162.33615; Time: 1605827178.75
Step 1898 of 3000; Error: 162.33606; Time: 1605827179.94
Step 1899 of 3000; Error: 162.33596; Time: 1605827181.17
Step 1900 of 3000; Error: 162.33586; Time: 1605827182.43
Step 1901 of 3000; Error: 162.33576; Time: 1605827183.63
Step 1902 of 3000; Error: 162.33566; Time: 1605827184.86
Step 1903 of 3000; Error: 162.33556; Time: 1605827186.17
Step 1904 of 3000; Error: 162.33546; Time: 1605827187.46
Step 1905 of 3000; Error: 162.3

Step 2032 of 3000; Error: 162.32291; Time: 1605827322.72
Step 2033 of 3000; Error: 162.32282; Time: 1605827323.99
Step 2034 of 3000; Error: 162.32272; Time: 1605827325.27
Step 2035 of 3000; Error: 162.32262; Time: 1605827326.43
Step 2036 of 3000; Error: 162.32252; Time: 1605827327.61
Step 2037 of 3000; Error: 162.32242; Time: 1605827328.92
Step 2038 of 3000; Error: 162.32233; Time: 1605827330.35
Step 2039 of 3000; Error: 162.32223; Time: 1605827331.77
Step 2040 of 3000; Error: 162.32213; Time: 1605827332.95
Step 2041 of 3000; Error: 162.32203; Time: 1605827333.96
Step 2042 of 3000; Error: 162.32194; Time: 1605827334.92
Step 2043 of 3000; Error: 162.32184; Time: 1605827335.98
Step 2044 of 3000; Error: 162.32174; Time: 1605827336.95
Step 2045 of 3000; Error: 162.32164; Time: 1605827337.85
Step 2046 of 3000; Error: 162.32154; Time: 1605827338.75
Step 2047 of 3000; Error: 162.32145; Time: 1605827339.67
Step 2048 of 3000; Error: 162.32135; Time: 1605827340.56
Step 2049 of 3000; Error: 162.3

Step 2176 of 3000; Error: 162.30886; Time: 1605827474.15
Step 2177 of 3000; Error: 162.30877; Time: 1605827475.32
Step 2178 of 3000; Error: 162.30867; Time: 1605827476.41
Step 2179 of 3000; Error: 162.30857; Time: 1605827477.53
Step 2180 of 3000; Error: 162.30847; Time: 1605827478.62
Step 2181 of 3000; Error: 162.30838; Time: 1605827479.70
Step 2182 of 3000; Error: 162.30828; Time: 1605827480.76
Step 2183 of 3000; Error: 162.30818; Time: 1605827481.82
Step 2184 of 3000; Error: 162.30808; Time: 1605827482.92
Step 2185 of 3000; Error: 162.30799; Time: 1605827484.04
Step 2186 of 3000; Error: 162.30789; Time: 1605827485.16
Step 2187 of 3000; Error: 162.30779; Time: 1605827486.29
Step 2188 of 3000; Error: 162.30770; Time: 1605827487.50
Step 2189 of 3000; Error: 162.30760; Time: 1605827488.67
Step 2190 of 3000; Error: 162.30750; Time: 1605827489.89
Step 2191 of 3000; Error: 162.30740; Time: 1605827491.13
Step 2192 of 3000; Error: 162.30731; Time: 1605827492.37
Step 2193 of 3000; Error: 162.3

Step 2320 of 3000; Error: 162.29486; Time: 1605827666.69
Step 2321 of 3000; Error: 162.29477; Time: 1605827667.98
Step 2322 of 3000; Error: 162.29467; Time: 1605827669.22
Step 2323 of 3000; Error: 162.29457; Time: 1605827670.47
Step 2324 of 3000; Error: 162.29448; Time: 1605827671.68
Step 2325 of 3000; Error: 162.29438; Time: 1605827672.83
Step 2326 of 3000; Error: 162.29428; Time: 1605827674.05
Step 2327 of 3000; Error: 162.29418; Time: 1605827675.26
Step 2328 of 3000; Error: 162.29409; Time: 1605827676.47
Step 2329 of 3000; Error: 162.29399; Time: 1605827677.62
Step 2330 of 3000; Error: 162.29389; Time: 1605827678.83
Step 2331 of 3000; Error: 162.29380; Time: 1605827680.00
Step 2332 of 3000; Error: 162.29370; Time: 1605827681.16
Step 2333 of 3000; Error: 162.29360; Time: 1605827682.39
Step 2334 of 3000; Error: 162.29350; Time: 1605827683.58
Step 2335 of 3000; Error: 162.29341; Time: 1605827684.88
Step 2336 of 3000; Error: 162.29331; Time: 1605827686.13
Step 2337 of 3000; Error: 162.2

Step 2464 of 3000; Error: 162.28090; Time: 1605827830.01
Step 2465 of 3000; Error: 162.28080; Time: 1605827831.29
Step 2466 of 3000; Error: 162.28071; Time: 1605827832.59
Step 2467 of 3000; Error: 162.28061; Time: 1605827833.81
Step 2468 of 3000; Error: 162.28051; Time: 1605827835.10
Step 2469 of 3000; Error: 162.28042; Time: 1605827836.37
Step 2470 of 3000; Error: 162.28032; Time: 1605827837.60
Step 2471 of 3000; Error: 162.28022; Time: 1605827838.90
Step 2472 of 3000; Error: 162.28013; Time: 1605827839.99
Step 2473 of 3000; Error: 162.28003; Time: 1605827841.19
Step 2474 of 3000; Error: 162.27993; Time: 1605827842.93
Step 2475 of 3000; Error: 162.27984; Time: 1605827844.53
Step 2476 of 3000; Error: 162.27974; Time: 1605827845.70
Step 2477 of 3000; Error: 162.27964; Time: 1605827847.16
Step 2478 of 3000; Error: 162.27955; Time: 1605827848.38
Step 2479 of 3000; Error: 162.27945; Time: 1605827849.51
Step 2480 of 3000; Error: 162.27935; Time: 1605827850.86
Step 2481 of 3000; Error: 162.2

Step 2608 of 3000; Error: 162.26697; Time: 1605828043.73
Step 2609 of 3000; Error: 162.26688; Time: 1605828047.41
Step 2610 of 3000; Error: 162.26678; Time: 1605828050.30
Step 2611 of 3000; Error: 162.26668; Time: 1605828053.81
Step 2612 of 3000; Error: 162.26659; Time: 1605828056.71
Step 2613 of 3000; Error: 162.26649; Time: 1605828059.13
Step 2614 of 3000; Error: 162.26639; Time: 1605828061.06
Step 2615 of 3000; Error: 162.26630; Time: 1605828063.33
Step 2616 of 3000; Error: 162.26620; Time: 1605828065.41
Step 2617 of 3000; Error: 162.26610; Time: 1605828067.58
Step 2618 of 3000; Error: 162.26601; Time: 1605828069.24
Step 2619 of 3000; Error: 162.26591; Time: 1605828070.85
Step 2620 of 3000; Error: 162.26581; Time: 1605828072.35
Step 2621 of 3000; Error: 162.26572; Time: 1605828073.93
Step 2622 of 3000; Error: 162.26562; Time: 1605828075.43
Step 2623 of 3000; Error: 162.26552; Time: 1605828077.03
Step 2624 of 3000; Error: 162.26543; Time: 1605828078.59
Step 2625 of 3000; Error: 162.2

Step 2752 of 3000; Error: 162.25308; Time: 1605828264.21
Step 2753 of 3000; Error: 162.25298; Time: 1605828265.54
Step 2754 of 3000; Error: 162.25288; Time: 1605828266.84
Step 2755 of 3000; Error: 162.25279; Time: 1605828268.27
Step 2756 of 3000; Error: 162.25269; Time: 1605828269.74
Step 2757 of 3000; Error: 162.25259; Time: 1605828271.26
Step 2758 of 3000; Error: 162.25250; Time: 1605828272.82
Step 2759 of 3000; Error: 162.25240; Time: 1605828274.45
Step 2760 of 3000; Error: 162.25231; Time: 1605828276.02
Step 2761 of 3000; Error: 162.25221; Time: 1605828277.80
Step 2762 of 3000; Error: 162.25211; Time: 1605828279.99
Step 2763 of 3000; Error: 162.25202; Time: 1605828282.15
Step 2764 of 3000; Error: 162.25192; Time: 1605828283.73
Step 2765 of 3000; Error: 162.25182; Time: 1605828285.53
Step 2766 of 3000; Error: 162.25173; Time: 1605828287.56
Step 2767 of 3000; Error: 162.25163; Time: 1605828289.25
Step 2768 of 3000; Error: 162.25153; Time: 1605828291.34
Step 2769 of 3000; Error: 162.2

Step 2896 of 3000; Error: 162.23921; Time: 1605828493.39
Step 2897 of 3000; Error: 162.23912; Time: 1605828494.78
Step 2898 of 3000; Error: 162.23902; Time: 1605828496.20
Step 2899 of 3000; Error: 162.23892; Time: 1605828497.62
Step 2900 of 3000; Error: 162.23883; Time: 1605828499.27
Step 2901 of 3000; Error: 162.23873; Time: 1605828500.71
Step 2902 of 3000; Error: 162.23863; Time: 1605828502.22
Step 2903 of 3000; Error: 162.23854; Time: 1605828503.56
Step 2904 of 3000; Error: 162.23844; Time: 1605828504.83
Step 2905 of 3000; Error: 162.23835; Time: 1605828506.09
Step 2906 of 3000; Error: 162.23825; Time: 1605828507.33
Step 2907 of 3000; Error: 162.23815; Time: 1605828508.74
Step 2908 of 3000; Error: 162.23806; Time: 1605828510.29
Step 2909 of 3000; Error: 162.23796; Time: 1605828511.69
Step 2910 of 3000; Error: 162.23787; Time: 1605828513.05
Step 2911 of 3000; Error: 162.23777; Time: 1605828514.63
Step 2912 of 3000; Error: 162.23767; Time: 1605828516.06
Step 2913 of 3000; Error: 162.2

In [ ]:
outP = open("rr_p.csv", "w")
outQ = open("rr_q.csv", "w")
np.savetxt(outP, fP, delimiter=',', fmt='%1.4f')
np.savetxt(outQ, fQ, delimiter=',', fmt='%1.4f')

##### Predicted Ratings
This computes the predicted rating for each user and item. 

In [ ]:
Preds = np.dot(fP,fQ.T)

In [ ]:
outPreds = open("runway_preds.csv", "w")
np.savetxt(outPreds, Preds, delimiter=',', fmt='%1.4f')

##### Mean Asbolute Error
Lastly we calculate the Mean Absolute Error (MAE) for each user as well as the overall MAE for the disctict users and items. 

In [11]:
totCount = 0
totError = 0

for u in range(M):
    err_u = 0
    rateCount_u = 0
    for j in range(N):
        if (R1[u,j] > 0): ### Only use known ratings computing error
            rateCount_u += 1
            err_u += abs(np.dot(fP[u],fQ[j]) - R1[u,j])
    print("Mean Absolute Error for User %d = %0.3f" %(u, err_u/rateCount_u))
    totCount += rateCount_u
    totError += err_u
print
print("Overall Mean Absolute Error = %0.3f" %(totError/totCount))

Mean Absolute Error for User 0 = 0.010
Mean Absolute Error for User 1 = 0.010
Mean Absolute Error for User 2 = 0.010
Mean Absolute Error for User 3 = 0.009
Mean Absolute Error for User 4 = 0.010
Mean Absolute Error for User 5 = 0.010
Mean Absolute Error for User 6 = 0.010
Mean Absolute Error for User 7 = 0.010
Mean Absolute Error for User 8 = 0.010
Mean Absolute Error for User 9 = 0.010
Mean Absolute Error for User 10 = 0.009
Mean Absolute Error for User 11 = 0.010
Mean Absolute Error for User 12 = 0.010
Mean Absolute Error for User 13 = 0.010
Mean Absolute Error for User 14 = 0.010
Mean Absolute Error for User 15 = 0.009
Mean Absolute Error for User 16 = 0.010
Mean Absolute Error for User 17 = 0.010
Mean Absolute Error for User 18 = 0.010
Mean Absolute Error for User 19 = 0.010
Mean Absolute Error for User 20 = 0.009
Mean Absolute Error for User 21 = 0.010
Mean Absolute Error for User 22 = 0.010
Mean Absolute Error for User 23 = 0.010
Mean Absolute Error for User 24 = 0.009
Mean Absol

Mean Absolute Error for User 252 = 0.009
Mean Absolute Error for User 253 = 0.010
Mean Absolute Error for User 254 = 0.010
Mean Absolute Error for User 255 = 0.009
Mean Absolute Error for User 256 = 0.010
Mean Absolute Error for User 257 = 0.010
Mean Absolute Error for User 258 = 0.010
Mean Absolute Error for User 259 = 0.010
Mean Absolute Error for User 260 = 0.010
Mean Absolute Error for User 261 = 0.010
Mean Absolute Error for User 262 = 0.010
Mean Absolute Error for User 263 = 0.010
Mean Absolute Error for User 264 = 0.010
Mean Absolute Error for User 265 = 0.010
Mean Absolute Error for User 266 = 0.010
Mean Absolute Error for User 267 = 0.010
Mean Absolute Error for User 268 = 0.010
Mean Absolute Error for User 269 = 0.010
Mean Absolute Error for User 270 = 0.010
Mean Absolute Error for User 271 = 0.010
Mean Absolute Error for User 272 = 0.010
Mean Absolute Error for User 273 = 0.010
Mean Absolute Error for User 274 = 0.010
Mean Absolute Error for User 275 = 0.010
Mean Absolute Er

Mean Absolute Error for User 533 = 0.010
Mean Absolute Error for User 534 = 0.010
Mean Absolute Error for User 535 = 0.010
Mean Absolute Error for User 536 = 0.009
Mean Absolute Error for User 537 = 0.010
Mean Absolute Error for User 538 = 0.009
Mean Absolute Error for User 539 = 0.010
Mean Absolute Error for User 540 = 0.010
Mean Absolute Error for User 541 = 0.010
Mean Absolute Error for User 542 = 0.010
Mean Absolute Error for User 543 = 0.009
Mean Absolute Error for User 544 = 0.010
Mean Absolute Error for User 545 = 0.010
Mean Absolute Error for User 546 = 0.010
Mean Absolute Error for User 547 = 0.009
Mean Absolute Error for User 548 = 0.010
Mean Absolute Error for User 549 = 0.010
Mean Absolute Error for User 550 = 0.010
Mean Absolute Error for User 551 = 0.010
Mean Absolute Error for User 552 = 0.010
Mean Absolute Error for User 553 = 0.010
Mean Absolute Error for User 554 = 0.010
Mean Absolute Error for User 555 = 0.010
Mean Absolute Error for User 556 = 0.009
Mean Absolute Er

Mean Absolute Error for User 798 = 0.010
Mean Absolute Error for User 799 = 0.010
Mean Absolute Error for User 800 = 0.010
Mean Absolute Error for User 801 = 0.010
Mean Absolute Error for User 802 = 0.010
Mean Absolute Error for User 803 = 0.010
Mean Absolute Error for User 804 = 0.010
Mean Absolute Error for User 805 = 0.010
Mean Absolute Error for User 806 = 0.010
Mean Absolute Error for User 807 = 0.010
Mean Absolute Error for User 808 = 0.010
Mean Absolute Error for User 809 = 0.007
Mean Absolute Error for User 810 = 0.010
Mean Absolute Error for User 811 = 0.010
Mean Absolute Error for User 812 = 0.010
Mean Absolute Error for User 813 = 0.009
Mean Absolute Error for User 814 = 0.009
Mean Absolute Error for User 815 = 0.010
Mean Absolute Error for User 816 = 0.010
Mean Absolute Error for User 817 = 0.010
Mean Absolute Error for User 818 = 0.010
Mean Absolute Error for User 819 = 0.010
Mean Absolute Error for User 820 = 0.009
Mean Absolute Error for User 821 = 0.010
Mean Absolute Er

The Overall MAE using distinct users (~900) was .010 while the overall MAE for 2500 users was .018. The MAE is pretty small and indicates that the matrix factorization performed well using stochastic gradient descent. 